[New rule] disallow `aria-hidden="true"` on focusable elements

khiga8 opened the pull request:
Fixes: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/881

This adds a new rule that flags when `aria-hidden="true"` is set on focusable elements ~~or contain child elements that are focusable~~. People often set `aria-hidden="true"` to hide an element from AT users, but if the element is focusable, AT users can still access it which can be confusing.

I wasn't sure what the contributing process is like so wanted to get a draft PR going to get thoughts. I looked through the list of jsx-a11y rules and I couldn't find a rule for something like this but let me know if I missed it. If maintainers are okay moving forward with this, I will add a doc page. I've ran the lint and added a test.

Related references: 
- [aria-hidden elements do not contain focusable elements](https://dequeuniversity.com/rules/axe/html/4.4/aria-hidden-focus)
- [Element with aria-hidden has no content in sequential focus navigation](https://www.w3.org/WAI/standards-guidelines/act/rules/6cfa84/proposed/)
- [MDN aria-hidden](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-hidden) warning: `Do not use aria-hidden="true" on focusable elements.`

ljharb commented on src/rules/no-aria-hidden-on-focusable.js:
```diff
@@ -0,0 +1,65 @@
+  create: (context) => {
+    const elementType = getElementType(context);
+    return {
+      JSXOpeningElement: (node) => {
```
```suggestion
      JSXOpeningElement(node) {
```

ljharb commented on src/rules/no-aria-hidden-on-focusable.js:
```diff
@@ -0,0 +1,65 @@
+    schema: [schema],
+  },
+
+  create: (context) => {
```
```suggestion
  create(context) {
```

ljharb commented on __tests__/src/rules/no-aria-hidden-on-focusable-test.js:
```diff
@@ -0,0 +1,41 @@
+ruleTester.run('no-aria-hidden-on-focusable', rule, {
+  valid: [
+    { code: '<div aria-hidden="true" />;' },
+    // { code: '<img aria-hidden="true" />' },
+    // { code: '<a aria-hidden="false" href="#" />' },
+    // { code: '<button tabIndex="-1" />' },
+    // { code: '<button />' },
+    // { code: '<a href="/" />' },
+  ].map(parserOptionsMapper),
+  invalid: [
+    // { code: '<div aria-hidden="true" tabIndex="0" />;', errors: [expectedError] },
+    // { code: '<input aria-hidden="true" />;', errors: [expectedError] },
+    // { code: '<a href="/" aria-hidden="true" />', errors: [expectedError] },
+    // { code: '<button aria-hidden="true" />', errors: [expectedError] },
+    // { code: '<textarea aria-hidden="true" />', errors: [expectedError] },
```
can these be uncommented?

khiga8 replied:
Ah I was trying to debug a test. sorry about that.

ljharb reviewed (commented):
Seems good overall.

khiga8 commented:
Added more test cases and fixed a bug in [bc220ad](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/pull/888/commits/bc220adf8efe3f594d1375573b84f0c64fcf2b9a).

Added a doc page in [e96c9db](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/pull/888/commits/e96c9db87f52d17151c7718202c256e871b3202d).

ljharb commented on src/rules/no-aria-hidden-on-focusable.js:
```diff
@@ -0,0 +1,66 @@
+/* Returns true when element is interactive and does not have `tabIndex=-1`, or when an element has a tabIndex that isn't -1. */
+function isFocusable(type, attributes) {
+  const tabIndex = getTabIndex(getProp(attributes, 'tabIndex'));
+  return (tabIndex !== -1 && isInteractiveElement(type, attributes)) || (tabIndex !== undefined && tabIndex !== -1);
```
```suggestion
  return tabIndex !== -1 && (
    tabIndex !== undefined
    || isInteractiveElement(type, attributes)
  );
```

ljharb commented on src/rules/no-aria-hidden-on-focusable.js:
```diff
@@ -0,0 +1,66 @@
+          });
+        }
+        const { children } = node.parent;
+        children.forEach((child) => {
+          if (child.type === 'JSXElement' && child.openingElement) {
+            const openingChildElement = child.openingElement;
+            const childType = elementType(openingChildElement);
+            const childAttributes = openingChildElement.attributes;
+            if (isFocusable(childType, childAttributes)) {
+              context.report({
+                node: openingChildElement,
+                message: errorMessage,
+              });
+            }
+          }
+        });
```
```suggestion
        children.filter((child) => {
          if (child.type !== 'JSXElement' || !child.openingElement) {
            return false;
          }
          const openingChildElement = child.openingElement;
          const childType = elementType(openingChildElement);
          const childAttributes = openingChildElement.attributes;
          return isFocusable(childType, childAttributes);
        }).forEach((child) => {
          context.report({
            node: child.openingElement,
            message: errorMessage,
          });
        });
```

jessebeach commented on src/rules/no-aria-hidden-on-focusable.js:
```diff
@@ -0,0 +1,70 @@
+            message: errorMessage,
+          });
+        }
+        const { children } = node.parent;
```
Generally, you want to use ESLint's visitor pattern to visit nodes, rather than doing your own tree traversal within a rule down into child nodes. What is purpose of creating reports for focusable children of the opening node?

khiga8 replied:
Thank you for pointing this out. I'm new to ESLint rule creation so uh.. please excuse my non-idiomatic code 🙇‍♀️.

My intention had been to cover cases where `aria-hidden` is set on an element that contains focusable children like:
```.jsx
<div aria-hidden="true">
   <a href="/">Some site</a>
</div>
```

Is this something that can be accomplished with the visitor pattern? I'll have to do a bit more digging into ESLint conventions.

jessebeach replied:
> My intention had been to cover cases where aria-hidden is set on an element that contains focusable children like:

Ok, I understand. For this rule, I'm going to recommend that you keep it narrowly focused to just focusable elements and not their ancestors or descendants. The problem of the cascading effects of aria-hidden on the AX tree is a thorny one and it will complicate a rule that is otherwise straightforward and easy to reason about. Plus, these sorts of tree relationships are hard to lint for, because you're only getting a snippet of the total render tree while linting, so that incidence of positive cases is going to be low with this static linting method. The reason I really like your rule is that it can be narrowly focused to focusable elements and there is good literature about how `aria-hidden` should interact with these.

jessebeach replied:
> please excuse my non-idiomatic code 🙇‍♀️

No need for apologies :)

jessebeach reviewed (changes requested):
@khiga8 this is a great idea for a rule. You're off to a solid start. I'd love to move your `isFocusable` function to its own utility. And I have a question about the reason behind traversing child elements. 

I think you can get this to a place where we can get it committed. I look forward to the updates!

jessebeach reviewed (approved):
This is amazing @khiga8! Thank you so much for proposing this rule and going through the iterations to develop it. Let's get it landed!

jessebeach commented:
@ljharb ready to land
